design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,129 @@
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will become a persistent query instead of just a metadata operation like 
+`CREATE TABLE`, `CREATE STREAM`, or `CREATE SOURCE STREAM`. Source table will materialize the input topic data as 
+a RocksDB instance for interactive queries in the KsqlDB servers.
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
+* `DESCRIBE ... EXTENDED` should show source streams and tables as read-only  
+* Add materialization of the `source table` to make it pull-query accessible
+* Make source stream/table immutable from KSQL by rejecting insertion requests
+ 
+
+## What is not in scope
+* We will not add role-based write permission for source stream/table in v1, even though it seems to be useful.
+
+## Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
+* createStream
+* createTable
+
+with the sample syntax as:
+```
+CREATE (OR REPLACE)? (SOURCE)? TABLE (IF NOT EXISTS)?...
+```
+Then inside `AstBuilder#visitCreateStream` and `AstBuilder#visitCreateTable`, the built classes for stream/table 
+instances would be `CreateStream` and `CreateTable`, which are the only two subclasses of `CreateSource`. It makes sense 
+to add the `SOURCE` as a flag into `CreateSource`, and let `AstBuilder` pass the flag into the struct 
+for later stage reference of read-only.
+
+Additionally, when user calls `DESCRIBE ... EXTENDED`, the read-only attribute shall be displayed for stream/table. 
+
+### Make Source Stream/Table Read-only
+We plan to add restriction check into the `DistributingExecutor#throwIfInsertOnReadOnlyTopic` or create a similar function, 
+which could verify whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+### Source Table Materialization
+In the call `EngineExecutor#execute`, we will check whether given plan has a `queryPlan` before deciding to proceed 
+creating a persistent query. To bypass this check, we need to provide a concrete query plan within the input plan. We would 
+inject a dummy query plan in `EngineExecutor#plan` phase, when the statement is a CreateTable statement with `SOURCE` tag.
+
+In addition, when user tries to show all the running queries, the materialization query should also be displayed as it takes 
+part of the computation resource. This would be a special type of running query without `SELECT` clause or a query id, as it 
+should be only associated with source table, and user could choose to `DROP` the entire table if they don't want to use resource 
+to do the materialization.
+
+Here comes a precaution we might be aware. The original `CREATE TABLE` command does not need materialization, so there is 
+no enforcement for primary key. However, in order to let pull query run on the materialized state, the store must be keyed so that 
```

Comment:
Are you sure about this? I thought that we do require that a `CREATE TABLE` has at least one primary key column - what we don't ensure is that the actual data that is sent matches the said schema (otherwise it can't be used for anything downstream really)

Reply:
That sounds right, I think primary key is required for CT